### PR TITLE
feat(NgTemplateOutlet): add NgTemplateOutlet directive

### LINKED ISSUE #7615: [feature] A directive to insert / stamp out TemplateRefs
There are cases where we've got a reference to a `TemplateRef` and would like to use it to stamp out live HTML fragments. Currently it can be easily done from the JS side (`viewContainerRef.createEmbeddedView(tplRef)`) but this issue is about creating a directive that would allow us to do this "inserting" from a template.

A "workish" example of such directive:

```javascript
@Directive({
    selector: 'ng-insert-tpl'
})
export class NgInsertTpl {
    constructor(private _viewContainerRef: ViewContainerRef) {
    }
    
    @Input() set template(tplRef: TemplateRef) {
        this._viewContainerRef.createEmbeddedView(tplRef);
    }
}
```

Naming things is hard so `ng-insert-tpl` is definitively up to debate.

This is part of https://github.com/angular/angular/issues/7546#issuecomment-196848167, it could be used in action like so: http://plnkr.co/edit/sleVVrgqo66JN9G48yyj?p=preview

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
- **Please check if the PR fulfills these requirements**
- [ x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [ x] Tests for the changes have been added (for bug fixes / features)
- [x ] Docs have been added / updated (for bug fixes / features)
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

New feature
- **What is the current behavior?** (You can also link to an open issue here)

There is no standard way of creating embeded views from a template.
- **What is the new behavior (if this is a feature change)?**

A new `NgInsert` directive makes it possible to create an embeded view from a provided `TemplateRef`
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No
- **Other information**:

Closes #7615

### CODE REVIEW

**Comment by pkozlowski-opensource** [comment]
@mhevery opening a PR to kick off detailed discussion on the NgInsert / NgOutlet directive discussed in https://github.com/angular/angular/issues/7615.

Opening a PR with the simplest possible and useful implementation. We can gradually add more features like:
* passing local variables
* ability to insert components, not only TemplateRef

cc @jelbourn

**Comment by jelbourn** [comment]
@pkozlowski-opensource Thanks for kicking this off! 

I think it's important to consider component loading in the initial API;  in particular, we may want a single property that can be set to load either a `TemplateRef` or a component (rather than always taking `TemplateRef`). I like the idea of some wrapper object (like how we have `Portal`) because it lets you add additional information/metadata to whatever you're loading (such as `locals` for a `TempateRef` or an origin `ElementRef` for a component).

File: modules/angular2/src/common/directives/ng_insert.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import {Directive, Input, ViewContainerRef, ViewRef, TemplateRef} from 'angular2/core';
+import {isPresent} from 'angular2/src/facade/lang';
+
+/**
+ * Creates and inserts an embedded view based on a prepared `TemplateRef`.
+ *
+ * ### Syntax
+ * - `<ng-insert [templateRef]="templateRefExpression"></ng-insert>`
```
Comment (mhevery) [inline modules/angular2/src/common/directives/ng_insert.ts]: I think a better syntax would be:
`<template [ngInsert]="templateRefExpression"></template>`

The reason is that we would prefer not have floating `<ng-insert>`  in the final DOM.

**Comment by mhevery** [comment]
Looks, good. Sorry, my only comment is about name (bit o bikeshedding.)

**Comment by pkozlowski-opensource** [comment]
@mhevery thnx for the review. I like the suggestion of having `<template [ngInsert]="exp">` syntax as indeed, it will avoid those "dangling" elements. Will change / update the PR. (my part of bikeshedding here: it still bothers me that we use `<template>` both as a _capture_ and _insertion_ points, but not sure why... Why I would like to see them as distinct concepts? I can't think of good reasons, but yeh,....)

@jelbourn yep, let's build on top of this for components, locals etc. I will be in MTV in 4 weeks so we can iterate on this one.

**Comment by mhevery** [comment]
@pkozlowski-opensource what is your thought on `ngOutlet` istead of `ngInsert`?

I am with you on using `template` for both input and output. But like you other than aesthetics not sure why it matters. Keep in mind that `template` is part of the web spec, so if we do come up with `template-output` we would be breaking away from the spec.

**Comment by mhevery** [comment]
we could do comments `<!--nd-directive: [ngOulter]="expression" -->`

**Comment by pkozlowski-opensource** [comment]
> what is your thought on ngOutlet istead of ngInsert?

I've "tested" `ngOutlet` on people in my office and it was too confusing with a router - this is why I've proposed NgInsert in this PR.

**Comment by jelbourn** [comment]
I think `<template>` is just fine for both. Adding behavior to comments strikes me as extremely unexpected behavior. 

I also like `ngOutlet` over `ngInsert`. If you're just doing to support `TemplateRef`, though, a more meaningful name would be `TemplateHost`.

**Comment by pkozlowski-opensource** [comment]
OK, did all the renamings / refactors discussed in the PR. Rebased, build is green.

**Comment by mhevery** [comment]
The PR looks great, but we got into a naming discussion. The issue is that `insert` implies that you are adding an item to a list, in this case the `TemplateRef`. But in our case what is being inserted is a `View` not a `TemplateRef`, it is just that we get the `View` from the `TemplateRef`, so we don't think we want `insert.

Sorry, I know this is bikeshedding, but naming is important.

The team thinks we should call it `ngTemplateOutlet` to be aligned with `routeOutlet` since they do a very similar things, one with `TemplateRef` the other with `Route`

**Comment by pkozlowski-opensource** [comment]
Personally I don't feel too strong about names so I'm perfectly fine with `ngTemplateOutlet`. Will update the PR accordingly.

Thnx for looking into this / discussing!

**Comment by pkozlowski-opensource** [comment]
@mhevery did the discussed renaming. Dart build is failing but I don't think it is related to my change.

**Comment by mhevery** [comment]
Yes, dart change is unrelated.

**Comment by bluEEil** [comment]
Hello, Im trying to use this feature but with no success. 
I tried following this example: [http://5thingsangular.github.io/2016/04/18/issue-2.html](5thingsAngular)
but with no succuess, no matter what I do I get an template parse error. I wrote the entire problem with more explanation on stack overflow as well:
[http://stackoverflow.com/questions/36965977/ngtemplateoutlet-doesnt-work-for-me](NgTemplateOutlet doesn't work for me)

I will really appreciate any help with this

**Pull request closed without merging**